### PR TITLE
feat: implement media track extraction

### PR DESCRIPTION
**Pull request by polw1**
## Summary

  Adds media track extraction.

## Changes

  - Adds `get_tracks` Tauri command and `getTracks` guest-js API
  - Extracts MP4 track details from `moov/trak` boxes
  - Returns video, audio, subtitle, and unknown track information

## Validation

- [x] npm run standards

### CODE REVIEW

File: crates/media-parser/src/format/mp4/atoms/media.rs [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+   pub sample_rate: Option<u32>,
+   pub entry_count: u32,
+}
+
+const TKHD_V0_TRACK_ID_OFFSET: usize = 12;
+const TKHD_V0_DURATION_OFFSET: usize = 20;
+const TKHD_V0_WIDTH_OFFSET: usize = 76;
+const TKHD_V0_HEIGHT_OFFSET: usize = 80;
+const TKHD_V1_TRACK_ID_OFFSET: usize = 20;
+const TKHD_V1_DURATION_OFFSET: usize = 32;
```
Comment (velocitysystems) [inline crates/media-parser/src/format/mp4/atoms/media.rs]: `TKHD_V1_DURATION_OFFSET = 32` reads 4 bytes into the duration field. In a v1 tkhd payload this should be 28, not 32. Every file with a v1 tkhd (long files, and several encoders that always emit v1) will report a _corrupt_ track duration.

Currently, this isn't caught because the only fixture is a v0 audio file. Can we add a v1 unit test for `parse_tkhd` to validate this?

File: crates/media-parser/src/format/mp4/tracks.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+      &moov_data
+   };
+
+   let mut tracks = Vec::new();
+   for (fourcc, trak) in iter_boxes(moov_payload) {
+      if &fourcc != b"trak" {
+         continue;
+      }
+
+      tracks.push(parse_trak(trak)?);
```
Comment (velocitysystems) [inline crates/media-parser/src/format/mp4/tracks.rs]: `parse_trak(trak)?` propagates an error, so a single track missing tkhd/mdia/mdhd returns _zero_ tracks for the whole file . Should we instead skip and continue (optionally logging) rather than failing here?

File: crates/media-parser/src/format/mp4/atoms/media.rs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+pub fn parse_hdlr(hdlr: &[u8]) -> Option<[u8; 4]> {
+   hdlr
+      .get(HDLR_HANDLER_TYPE_OFFSET..HDLR_HANDLER_TYPE_OFFSET + 4)?
+      .try_into()
+      .ok()
+}
+
+pub fn parse_stsd(stsd: &[u8]) -> Option<SampleDescription> {
```
Comment (velocitysystems) [inline crates/media-parser/src/format/mp4/atoms/media.rs]: `width`/`height` and `channels`/`sample_rate` are read from the same payload offsets regardless of whether the entry is _audio_ or _video_, so e.g. an mp4a entry gets a spurious `width: Some(16)`.

File: guest-js/index.ts [hunk 4]
```diff
@@ -38,6 +38,23 @@ export interface MetadataOptions {
    headers?: Record<string, string>;
 }
 
+/**
+ * Track information extracted from a media file.
+ */
+export interface TrackInfo {
```
Comment (velocitysystems) [inline guest-js/index.ts]: All variant fields are _optional_ for every kind, so `if (track.kind === 'video') { track.width }` gives no narrowing (width stays `number | undefined`). A union keyed on _kind_ (video → width/height, audio → channels/sampleRate) would let the compiler guarantee presence per kind, matching the Rust `TrackType` enum.

As an example. we use discriminated unions on the TS types in the [tauri-plugin-download](https://github.com/silvermine/tauri-plugin-download) repository.

**Review by velocitysystems — CHANGES_REQUESTED** [review]

File: crates/media-parser/src/format/mp4/atoms/media.rs (hunk 1 above)
Comment (polw1) [reply crates/media-parser/src/format/mp4/atoms/media.rs]: Good catch on tkhd v1 thx

File: crates/media-parser/src/format/mp4/tracks.rs (hunk 2 above)
Comment (polw1) [reply crates/media-parser/src/format/mp4/tracks.rs]: Since these boxes are required by the spec, propagating the error felt reasonable,

That said, I have no objection to switching to a best-effort approach and skipping invalid tracks,
Thank you

File: guest-js/index.ts (hunk 4 above)
Comment (polw1) [reply guest-js/index.ts]: Agreed, and thank you for providing this example

File: crates/media-parser/src/format/mp4/tracks.rs [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+               let (channels, sample_rate) = audio_params(payload);
+               SampleEntry::Audio {
+                  channels,
+                  sample_rate,
+               }
+            }
+            _ => SampleEntry::None,
+         })
+      });
+   let sample_durations = stbl
```
Comment (velocitysystems) [inline crates/media-parser/src/format/mp4/tracks.rs]: Expanded sample tables are computed for _every_ track, then discarded. Track count is unbounded so a crafted file can drive `get_tracks` to tens of GB of transient allocation → OOM.

Can we avoid the DoS vulnerability and the unused sample tables?

File: crates/media-parser/src/types.rs [hunk 6]
```diff
@@ -78,19 +78,19 @@ pub struct AudioTrackMeta {
...
 /// Unknown/other track metadata.
-#[derive(Debug, Clone, PartialEq)]
+#[derive(Debug, Clone, PartialEq, Serialize)]
 pub struct UnknownTrackMeta {
    pub base: BaseTrackMeta,
 }
 
 /// Discriminated union of track kinds.
-#[derive(Debug, Clone, PartialEq)]
+#[derive(Debug, Clone, PartialEq, Serialize)]
```
Comment (velocitysystems) [inline crates/media-parser/src/types.rs]: The IPC boundary serializes the flat `TrackInfo` DTO, but this new derive makes `TrackType` itself serialize as an externally-tagged enum which doesn't match the documented TS TrackInfo union.

File: crates/media-parser/Cargo.toml [hunk 7]
```diff
@@ -12,6 +12,7 @@ tokio = { version = "1.48.0", features = ["fs", "io-util", "macros", "rt-multi-t
 reqwest = { version = "0.13.2", features = ["stream"] }
 bytes = "1.10.1"
 futures = "0.3.31"
+log = "0.4.33"
```
Comment (velocitysystems) [inline crates/media-parser/Cargo.toml]: For Silvermine projects we use the `tracing` crate.

File: crates/media-parser/src/format/mp4/tracks.rs
Comment (velocitysystems) [inline crates/media-parser/src/format/mp4/tracks.rs:1]: `tracks.rs` has no inline tests, so per-kind dispatch, skip-malformed-trak, and `classify_handler` are only exercised by one happy-path audio fixture. Should we add tests?

**Review by velocitysystems — CHANGES_REQUESTED** [review]

File: crates/media-parser/src/format/mp4/tracks.rs (hunk 5 above)
Comment (polw1) [reply crates/media-parser/src/format/mp4/tracks.rs]: Good point. I’ll avoid expanding `stts`/`stsz` in `get_tracks`.

Its better to keep sample-table data compact for track metadata, expose only aggregate fields like `sample_count`, `stts_entry_count`, `fixed_sample_size`, and `stsz_sample_count`, and reserve per-sample expansion for a future API that explicitly needs it with its own limits.

File: crates/media-parser/src/format/mp4/atoms/media.rs [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+   Some(TrackHeader {
+      id: read_u32_be(tkhd, track_id_offset)?,
+      duration,
+      width: read_fixed_16_16(tkhd, width_offset).unwrap_or(0),
+      height: read_fixed_16_16(tkhd, height_offset).unwrap_or(0),
+   })
+}
+
+pub fn parse_mdhd(mdhd: &[u8]) -> Option<MediaHeader> {
```
Comment (velocitysystems) [inline crates/media-parser/src/format/mp4/atoms/media.rs:86]: `parse_mdhd` has the same version-dependent offsets that caused the v1 tkhd bug, but no direct test — only the v0 integration fixture covers it. Could we add v0/v1 unit tests like the tkhd ones?

**Review by velocitysystems — CHANGES_REQUESTED** [review]

File: crates/media-parser/src/format/mp4/atoms/media.rs [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+pub fn parse_hdlr(hdlr: &[u8]) -> Option<[u8; 4]> {
+   hdlr
+      .get(HDLR_HANDLER_TYPE_OFFSET..HDLR_HANDLER_TYPE_OFFSET + 4)?
+      .try_into()
+      .ok()
+}
+
+pub fn parse_stsd<T>(
```
Comment (velocitysystems) [inline crates/media-parser/src/format/mp4/atoms/media.rs:122]: `read_box` checks bounds with `offset + total_size > data.len()` (unchecked add), which `parse_stsd` now reaches with potentially untrusted stsd bytes.

An extended-size entry (size32 == 1) with a near-u64::MAX size wraps the add, bypasses the check, and panics — crashing the app if a malicious file is provided.

Should we use `offset.checked_add(total_size)`? in `read_box`?

File: crates/media-parser/src/format/mp3/mod.rs [hunk 10]
```diff
@@ -58,14 +59,64 @@ fn parse(reader: &dyn StreamReader) -> Pin<Box<dyn Future<Output = Result<Metada
...
+   parse as AsyncParser,
+   parse_tracks as AsyncTrackParser,
+);
 
 /// Main parsing function.
 async fn parse_mp3(reader: &dyn StreamReader) -> Result<Metadata> {
    metadata::read_metadata(reader).await
 }
 
+async fn read_tracks(reader: &dyn StreamReader) -> Result<Vec<TrackType>> {
```
Comment (velocitysystems) [inline crates/media-parser/src/format/mp3/mod.rs:80]: Should we add tests for `read_tracks`? (Incl channel-mode→channels mapping, the empty/invalid-header branches).

**Review by velocitysystems — CHANGES_REQUESTED** [review]

File: crates/media-parser/src/format/mp4/atoms/media.rs (hunk 9 above)
Comment (polw1) [reply crates/media-parser/src/format/mp4/atoms/media.rs:122]: agreed,
along with this fix, reviewed similar offset and size calculations in external-data parsing paths and hardened the relevant ones against overflow

File: crates/media-parser/src/format/mp3/mod.rs (hunk 10 above)
Comment (polw1) [reply crates/media-parser/src/format/mp3/mod.rs:80]: Good point. It also made me revisit a few related paths around external reads and result handling, so I added coverage for the read_tracks branches and tightened that behavior as well.

File: crates/media-parser/src/format/mp3/mod.rs [hunk 11]
```diff
@@ -58,14 +59,58 @@ fn parse(reader: &dyn StreamReader) -> Pin<Box<dyn Future<Output = Result<Metada
...
+   let (header, offset) = match frame::find_first_frame(reader, 0, frame::MAX_SYNC_SEARCH).await? {
```
Comment (jjhafer) [inline crates/media-parser/src/format/mp3/mod.rs]: `read_tracks` scans from offset `0` here (and passes `id3_size = 0` to `calculate_duration` on line 88), so unlike the metadata path — which computes `id3_end = 10 + tag_size` in `metadata.rs` and scans from there — it never skips the ID3v2 tag. Since `MAX_SYNC_SEARCH` is 64 KiB, an MP3 whose ID3v2 tag exceeds 64 KiB (common with embedded cover art) will have its first real frame fall outside the search window, so `find_first_frame` returns `NotFound` and `getTracks` returns `[]` — even though `getMetadata` reports the correct duration on the same file. Could we read the ID3 size once and start both scans there? That would also drop the double first-frame scan (this call plus the one inside `calculate_duration`).

*Comment added by AI model Claude*

File: crates/media-parser/src/format/mp3/mod.rs [hunk 12]
```diff
@@ -58,14 +59,58 @@ fn parse(reader: &dyn StreamReader) -> Pin<Box<dyn Future<Output = Result<Metada
...
+      frame::FrameParseResult::NotFound | frame::FrameParseResult::EndOfData => {
+         return Ok(Vec::new());
+      }
+   };
+
+   let duration = duration::calculate_duration(reader, 0).await?;
+   let mut properties = HashMap::new();
+   properties.insert("offset".to_string(), offset.to_string());
+   properties.insert("bitrate_kbps".to_string(), header.bitrate_kbps.to_string());
+   properties.insert("mpeg_version".to_string(), format!("{:?}", header.version));
```
Comment (jjhafer) [inline crates/media-parser/src/format/mp3/mod.rs]: These three properties use `format!("{:?}", …)` on internal enums (`mpeg_version`, `mpeg_layer`, `duration_method`), which bakes Rust `Debug` output into the IPC wire contract. `Debug` isn't a stable representation — renaming an enum variant would silently change the JSON a JS consumer sees, with no compiler or test signal, while everything else in the bag uses `Display`/`to_string()`. What do you think about giving these an explicit `Display`/match mapping too?

*Comment added by AI model Claude*

File: crates/media-parser/src/format/mp4/tracks.rs [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+      channels: Option<u16>,
+      sample_rate: Option<u32>,
+   },
+   None,
+}
+
+/// Reads all MP4 tracks from the `moov/trak` boxes.
+pub async fn read_tracks(reader: &dyn StreamReader) -> Result<Vec<TrackType>> {
+   let moov_data = find_and_read_moov_box(reader).await?;
+   let moov_payload = if moov_data.len() >= 8 && &moov_data[4..8] == b"moov" {
```
Comment (jjhafer) [inline crates/media-parser/src/format/mp4/tracks.rs]: The `moov` fourcc sits at bytes `[4..8]` for both an 8-byte header and a 16-byte 64-bit-extended-size header (`size32 == 1`), so this branch is taken in both cases but only strips 8 bytes. For a 64-bit-sized `moov` the leftover 8 bytes of the size field get fed to `iter_boxes`, which then finds no valid `trak` and returns `[]` with no error. Rare in practice — could we derive the header length via `read_box_header(&moov_data, 0)` and skip `header_len` rather than hard-coding 8?

*Comment added by AI model Claude*

File: crates/media-parser/src/format/mp4/tracks.rs [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+      if &fourcc != b"trak" {
+         continue;
+      }
+
+      // Best-effort: skip a malformed trak (missing the spec-required
+      // tkhd/mdia/mdhd) instead of failing the whole file, so one bad track
+      // doesn't drop every other track.
+      match parse_trak(trak) {
+         Ok(track) => tracks.push(track),
+         Err(e) => tracing::warn!("skipping malformed trak: {e}"),
```
Comment (jjhafer) [inline crates/media-parser/src/format/mp4/tracks.rs]: Skipping a malformed `trak` with a `warn!` is the right resilience call. One consequence worth weighing: a `moov` where *every* `trak` fails to parse returns `Ok([])`, indistinguishable from a legitimately track-less container — and the warning only reaches logs the frontend never sees, while non-media / missing-`moov` files correctly `Err`. Would it be worth surfacing the skipped-track count (or erroring when `trak` boxes were present but none parsed) so a caller can tell 'corrupt' from 'genuinely empty' apart?

*Comment added by AI model Claude*

File: crates/media-parser/src/format/mp4/atoms/media.rs [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+pub fn visual_dimensions(payload: &[u8]) -> (Option<u32>, Option<u32>) {
+   (
+      read_u16_be(payload, VISUAL_WIDTH_OFFSET).map(u32::from),
+      read_u16_be(payload, VISUAL_HEIGHT_OFFSET).map(u32::from),
+   )
+}
+
+pub fn audio_params(payload: &[u8]) -> (Option<u16>, Option<u32>) {
```
Comment (jjhafer) [inline crates/media-parser/src/format/mp4/atoms/media.rs:143]: These offsets (`channelcount` @16, `samplerate` 16.16 @24) are correct for a version-0/1 `AudioSampleEntry`, covering the common AAC/MP3-at-≤48 kHz case. They're wrong for a QuickTime version-2 sound description, where the layout shifts and the sample rate is a 64-bit float at a different offset — an 88.2/96/176.4/192 kHz file would report a garbage `sampleRate` from `rate >> 16`. Could we read the 2-byte version at payload offset 8 and either handle v2 or return `None` for it, so we don't emit a silently-wrong value?

*Comment added by AI model Claude*

File: guest-js/types.ts [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+ * Fields common to every track kind.
+ */
+export interface BaseTrackInfo<K extends TrackKind> {
+   kind: K;
+   id: number;
+   codec: string;
+   language?: string;
+   timescale: number;
+   duration: number;
+   properties: Record<string, string>;
```
Comment (jjhafer) [inline guest-js/types.ts:86]: `properties` reaches JS as an untyped `Record<string, string>` with no documented keys, but the Rust side emits a fixed set (`handler_type`, `tkhd_duration`, `sample_entry_count`, `stts_entry_count`, `sample_count`, `fixed_sample_size`, `stsz_sample_count` for MP4; a different set for MP3) as stringified numbers — and the integration test already asserts `properties.get("sample_count") == "45"`, so consumers will lean on these too. Since this is now a public IPC surface, could we either document `properties` as an unstable/diagnostic bag or promote the stable numeric fields to typed properties? It's easy to freeze accidentally as-is.

*Comment added by AI model Claude*

File: crates/media-parser/tests/mp4_metadata.rs [hunk 17]
```diff
@@ -97,3 +97,41 @@ async fn test_mov_meta_ilst_values() {
...
+
+#[tokio::test]
+async fn test_mp4_tracks_extraction() {
+   let path = fixtures_dir().join("sample_metadata.mp4");
+   let reader = FileStreamReader::new(&path).expect("Failed to open MP4 fixture");
+   let parser = MediaParser::new(reader);
+
+   let tracks = parser.tracks().await.expect("Failed to parse MP4 tracks");
+
+   assert_eq!(tracks.len(), 1);
```
Comment (jjhafer) [inline crates/media-parser/tests/mp4_metadata.rs:109]: This is the only end-to-end MP4 track fixture, and it's a single *audio* track — so the video path (tkhd fixed-16.16 width/height, the visual `stsd` entry), the subtitle/unknown paths, and multi-track handling are exercised only by synthetic byte arrays. Those unit tests also tend to write a value at `VISUAL_WIDTH_OFFSET` and assert it reads back from the same constant, so a wrong offset would still pass. For a feature whose headline is video track extraction, would you be open to adding one real multi-track video `.mp4` fixture asserting known width/height (ideally with a v1 `tkhd`)? That's the one check that breaks a shared parser/test blind spot.

*Comment added by AI model Claude*

File: crates/media-parser/src/format/mp3/mod.rs (hunk 11 above)
Comment (polw1) [reply crates/media-parser/src/format/mp3/mod.rs]: Good finding, you’re right. I’ll use the ID3 end offset and avoid scanning for the first frame twice

File: crates/media-parser/src/format/mp3/mod.rs (hunk 12 above)
Comment (polw1) [reply crates/media-parser/src/format/mp3/mod.rs]: Good idea, thx

File: guest-js/types.ts (hunk 16 above)
Comment (polw1) [reply guest-js/types.ts:86]: Choose “document” for now, need to think about this a bit more, but good catch—thanks!

File: crates/media-parser/src/format/mp3/mod.rs [hunk 18]
```diff
@@ -58,14 +59,63 @@ fn parse(reader: &dyn StreamReader) -> Pin<Box<dyn Future<Output = Result<Metada
...
    metadata::read_metadata(reader).await
 }
 
+async fn read_tracks(reader: &dyn StreamReader) -> Result<Vec<TrackType>> {
+   let id3_end = metadata::read_id3_end(reader).await?;
+   let (header, offset) = match frame::find_first_frame_with_fallback_origin(
+      reader,
+      id3_end,
+      frame::MAX_SYNC_SEARCH,
+      0,
```
Comment (velocitysystems) [inline crates/media-parser/src/format/mp3/mod.rs]: Passing `fallback_origin: 0` here diverges from the metadata/duration path, which calls `find_first_frame(reader, id3_size, ..)` and so effectively passes `fallback_origin = id3_size` (`duration.rs`). Since the acceptance test in `frame.rs` is `frame_offset.saturating_sub(fallback_origin) > MIN_SEARCH_BEFORE_FALLBACK` (1024), any candidate past byte 1024 is accepted here without next-frame confirmation, while `calculate_duration` keeps scanning.

The result is that `getTracks` and `getMetadata` can report different durations for the same file.

Could we pass `id3_end` here to match `find_first_frame`'s own semantics? If the laxer behaviour is deliberate (the `read_tracks_keeps_single_frame_after_id3v2_tag` test depends on it), could we add a doc comment on `find_first_frame_with_fallback_origin` explaining what `fallback_origin` means and why the two callers differ? It's currently the only `pub(super)` fn here without one.

_Comment added by AI model Claude_

File: crates/media-parser/src/format/mp4/tracks.rs [hunk 19]
```diff
@@ -0,0 +1,358 @@
...
+   let moov_data = find_and_read_moov_box(reader).await?;
+   let moov_payload = read_box(&moov_data, 0)
+      .filter(|box_read| box_read.fourcc == *b"moov")
+      .map(|box_read| box_read.payload)
+      .ok_or_else(|| MediaParserError::InvalidFormat("invalid moov box".to_string()))?;
+
+   let mut tracks = Vec::new();
+   let mut trak_count = 0usize;
+   let mut malformed_count = 0usize;
+   for (fourcc, trak) in iter_boxes(moov_payload) {
```
Comment (velocitysystems) [inline crates/media-parser/src/format/mp4/tracks.rs:48]: This loop has no cap on track count, and `moov` is only bounded by `MAX_MOOV_SIZE` (100 MiB). A minimal valid `trak` (`tkhd` + `mdia{mdhd,hdlr}`) is ~160 bytes, so a crafted file yields a very large `Vec<TrackType>` that is then converted to `TrackInfo` and JSON-serialized across the IPC boundary to the webview.

_Comment added by AI model Claude_

File: crates/media-parser/src/format/mp4/tracks.rs [hunk 20]
```diff
@@ -0,0 +1,358 @@
...
+      channels: Option<u16>,
+      sample_rate: Option<u32>,
+   },
+   None,
+}
+
+/// Reads all MP4 tracks from the `moov/trak` boxes.
+pub async fn read_tracks(reader: &dyn StreamReader) -> Result<Vec<TrackType>> {
+   let moov_data = find_and_read_moov_box(reader).await?;
+   let moov_payload = read_box(&moov_data, 0)
```
Comment (velocitysystems) [inline crates/media-parser/src/format/mp4/tracks.rs]: Using `read_box` here is the right fix for the 64-bit extended-size `moov` header, and the `read_tracks_supports_extended_size_moov_header` test covers it.

It does leave the two `moov` readers disagreeing though: `mp4/metadata.rs` still does the hardcoded 8-byte strip:

```rust
let moov_payload = if moov_data.len() >= 8 && &moov_data[4..8] == b"moov" {
   &moov_data[8..]
} else {
   &moov_data
};
```

So on a file with an extended-size `moov`, `getTracks` now works and `getMetadata` feeds the leftover 8 bytes of the size field into `extract_mvhd_info`. Would it be worth extracting this `read_box`-based extraction into `atoms/moov.rs` and having both call it, so the two paths can't drift?

_Comment added by AI model Claude_

File: crates/media-parser/src/format/registry.rs [hunk 21]
```diff
@@ -54,6 +54,21 @@ pub async fn parse_metadata(reader: &dyn StreamReader) -> Result<Metadata> {
...
 }
 
+/// Parses track metadata by detecting format and dispatching to the appropriate parser.
+pub async fn parse_tracks(reader: &dyn StreamReader) -> Result<Vec<TrackType>> {
+   let mut header = [0u8; 32];
+   reader.read_at(0, &mut header).await?;
+
+   let format = detect_format(&header).ok_or_else(|| {
+      MediaParserError::InvalidFormat(format!(
+         "Could not detect format from header: {:02X?}",
```
Comment (velocitysystems) [inline crates/media-parser/src/format/registry.rs]: This differs from `parse_metadata` just above, which returns the generic `"Could not detect format from file header"`. Here the first 16 bytes of the file are formatted into the error, and plugin errors are serialized to the frontend via `to_string()`, so the bytes reach JS.

Concretely, `await getTracks('/Users/x/.ssh/id_ed25519')` rejects with a message containing 16 bytes of that file, for any file the process can open. It's a narrow read primitive, but it's new in this PR and the existing command doesn't do it.

Could we use the same generic message as `parse_metadata`, and put the hex behind `tracing::debug!` if it's useful for debugging?

_Comment added by AI model Claude_

File: crates/media-parser/Cargo.toml [hunk 22]
```diff
@@ -12,6 +12,7 @@ tokio = { version = "1.48.0", features = ["fs", "io-util", "macros", "rt-multi-t
 reqwest = { version = "0.13.2", features = ["stream"] }
 bytes = "1.10.1"
 futures = "0.3.31"
+tracing = "0.1.44"
```
Comment (velocitysystems) [inline crates/media-parser/Cargo.toml:15]: `Cargo.lock` doesn't appear to have been regenerated after adding this dependency. Could we commit the regenerated lockfile?

_Comment added by AI model Claude_

File: guest-js/types.ts [hunk 23]
```diff
@@ -0,0 +1,151 @@
+// ============================================================================
+// Metadata Types
+// ============================================================================
+
+/**
+ * Single extracted metadata item.
+ */
+export interface Meta {
+  /** Raw metadata key (e.g., "@nam" for MP4, "TIT2" for MP3). */
+  key: string;
```
Comment (velocitysystems) [inline guest-js/types.ts]: Nit: This new file is indented with 2 spaces throughout, while `guest-js/index.ts` and the rest of the repo use 3 per the Silvermine standard.

_Comment added by AI model Claude_

File: crates/media-parser/src/format/mp4/atoms/media.rs [hunk 24]
```diff
@@ -0,0 +1,344 @@
...
+      let parsed = parse_mdhd(&mdhd).unwrap();
+      assert_eq!(parsed.timescale, 48_000);
+      assert_eq!(parsed.duration, 5_000_000_000);
+      assert_eq!(parsed.language, Some(*b"eng"));
+   }
+
+   #[test]
+   fn test_parse_stsd_video_entry() {
+      let mut visual_payload = vec![0u8; 78];
+      visual_payload[VISUAL_WIDTH_OFFSET..VISUAL_WIDTH_OFFSET + 2]
```
Comment (velocitysystems) [inline crates/media-parser/src/format/mp4/atoms/media.rs]: This test writes at `VISUAL_WIDTH_OFFSET`/`VISUAL_HEIGHT_OFFSET` and then asserts the values read back through those same two constants, so changing them to (say) 20/22 would leave it passing. It's the one test in the PR with that property — the `tkhd`/`mdhd`/audio tests all deliberately use spec-derived literals and even carry a comment saying why, which is a nice touch.

The real fixtures only partially backstop it: in both `multitrack_video.mp4` and `tkhd_v1_video.mp4` the `tkhd` dimensions equal the `stsd` dimensions (160x90), and `parse_trak` does `width.unwrap_or(tkhd.width)` — so an out-of-bounds offset returns `None` and silently falls back to the same 160, and the integration test still passes.

Could we use the literals `24`/`26` here to match the sibling tests? That closes the gap in one line.

_Comment added by AI model Claude_

**Review by velocitysystems — CHANGES_REQUESTED** [review]

File: crates/media-parser/src/format/mp3/mod.rs (hunk 18 above)
Comment (polw1) [reply crates/media-parser/src/format/mp3/mod.rs]: Thanks for raising this. After giving it some more thought, I decided to remove the absolute-offset fallback and align track discovery with the metadata path. 

Both now start scanning after the ID3v2 tag and accept a candidate only when it is followed by a structurally compatible MPEG frame. This makes validation independent of the ID3v2 tag size, prevents isolated headers from being accepted as streams, and keeps frame acceptance consistent across both APIs. The search remains bounded to 64 KiB, and regression tests cover the shared behavior.

File: crates/media-parser/src/format/mp4/tracks.rs (hunk 19 above)
Comment (polw1) [reply crates/media-parser/src/format/mp4/tracks.rs:48]: I added a hard cap of 1,000 `trak` boxes. `read_tracks` now fails before parsing the 1,001st box, including malformed ones, which bounds both the result allocation and downstream IPC serialization

File: src/commands.rs [hunk 25]
```diff
@@ -19,20 +43,111 @@ pub(crate) async fn get_metadata(
...
-      parser.metadata().await.map_err(Into::into)
+      parser.tracks().await
+   })?;
+
+   Ok(tracks.into_iter().map(TrackInfo::from).collect())
+}
+
+#[derive(serde::Serialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct TrackInfo {
```
Comment (velocitysystems) [inline src/commands.rs:68]: `TrackInfo` lives inside the private commands module and isn't re-exported, so the declared return type of get_tracks can't be named by Rust consumers or seen by rustdoc.

File: src/commands.rs [hunk 26]
```diff
@@ -19,20 +43,111 @@ pub(crate) async fn get_metadata(
...
+            ..Self::from_base("audio", audio.base)
+         },
+         TrackType::Subtitle(subtitle) => Self::from_base("subtitle", subtitle.base),
+         TrackType::Unknown(unknown) => Self::from_base("unknown", unknown.base),
+      }
+   }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (velocitysystems) [inline src/commands.rs:125]: Nit: The `Video` and `Audio` arms, the `sample_rate → sampleRate` serde rename, and the four kind strings are the TS `TrackKind` contract, and none are pinned by a test. Should we add a serialization test for these?

**Review by velocitysystems — CHANGES_REQUESTED** [review]
Thanks @polw1. Just two remaining observations for your consideration.

File: src/commands.rs (hunk 25 above)
Comment (polw1) [reply src/commands.rs:68]: the project currently treats the Tauri commands as an internal IPC layer: both `get_tracks` and `get_metadata` are `pub(crate)` functions inside the private `commands` module.

The supported Rust-facing API of this crate is currently plugin registration, while the commands and their DTOs are exposed to the frontend through IPC. The public frontend contract is defined and exported from `guest-js/types.ts`.

Consumers that need lower-level Rust access can depend directly on the `media-parser` crate and use its public API, including `MediaParser::tracks()` and `media_parser::TrackType`.

Re-exporting `TrackInfo` alone would make the IPC DTO nameable, but it would not make `get_tracks` callable by downstream Rust crates. It would also make the IPC representation part of this crate’s supported public Rust API.

Is the intention to make the command layer and its DTOs part of the public Rust API? If so, I can move and expose `TrackInfo` accordingly. Otherwise, keeping it internal seems consistent with the current architecture.

Comment (velocitysystems) [reply src/commands.rs:68]: Thanks @polw1. Per your comments I guess this is moot for a type whose only consumer is the IPC serializer of a private command. Let's keep it as-is.

File: src/commands.rs (hunk 26 above)
Comment (polw1) [reply src/commands.rs:125]: good idea =)

**Review by velocitysystems — APPROVED** [review]
LGTM
